### PR TITLE
[ci][postmerge/3] actually trigger gap builds

### PR DESCRIPTION
**Pull request by can-anyscale**
Implement the function to trigger a blocked build on postmerge

- Also cache the _get_builds function which is used repeatedly and quite expensive

Test:
- CI

### CODE REVIEW

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 1]
```diff
@@ -55,8 +57,19 @@ def get_gap_commits(self) -> List[str]:
...
+        build = [
+            build
+            for build in self._get_builds()
+            if build["commit"] == commit and build["state"] == "blocked"
+        ][0]
+        self.buildkite.jobs().unblock_job(
+            self.buildkite_organization,
+            self.buildkite_pipeline,
+            build["number"],
+            build["jobs"][0]["id"],  # first job is the blocked job
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:69]: is it possible to check some thing that is more reliable? like the label or the step?

**Comment by can-anyscale** [comment]
unblock a step with a special label

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 2]
```diff
@@ -29,7 +30,7 @@ def __init__(
         self.repo_checkout = repo_checkout
         self.days_ago = days_ago
 
-    def run(self) -> List[str]:
+    def run(self) -> List[Optional[str]]:
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:33]: why is this returning a list of optional string?

can we just convert `None` to empty string and drop the `Optional`?

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 3]
```diff
@@ -60,9 +61,33 @@ def get_gap_commits(self) -> List[str]:
             .split("\n")
         )
 
-    def _trigger_build(self, commit: str) -> int:
-        # TODO(can): Implement this method
-        pass
+    def _find_blocked_build_and_job(
+        self, commit: str
+    ) -> Tuple[Optional[Dict], Optional[Dict]]:
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:66]: could you add a comment on what the return values are? or maybe create a simple class/TypedDict for the return value?

why is the build number a dict? and job id a dict too?

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 4]
```diff
@@ -29,14 +30,14 @@ def __init__(
         self.repo_checkout = repo_checkout
         self.days_ago = days_ago
 
-    def run(self) -> List[str]:
+    def run(self) -> Dict[str, Optional[str]]:
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:33]: update the doc comment?

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 5]
```diff
@@ -60,9 +61,33 @@ def get_gap_commits(self) -> List[str]:
...
+        return None, None
+
     def _trigger_build(self, commit: str) -> Optional[str]:
-        # TODO(can): Implement this method
-        pass
+        build, job = self._find_blocked_build_and_job(commit)
+        if not build or not job:
+            return None
+
+        self.buildkite.jobs().unblock_job(
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:85]: this is always returning `None`?

**Review by aslonnie — APPROVED** [review]

**Pull request merged**